x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/cache/TransportSearchableSnapshotsNodeCachesStatsAction.java

```diff
@@ -0,0 +1,270 @@
+    TransportSearchableSnapshotsNodeCachesStatsAction.NodeRequest,
+    TransportSearchableSnapshotsNodeCachesStatsAction.NodeCachesStatsResponse> {
+
+    public static final String ACTION_NAME = "cluster:admin/xpack/searchable_snapshots/cache/stats";
+
+    public static final ActionType<NodesCachesStatsResponse> TYPE = new ActionType<>(ACTION_NAME, NodesCachesStatsResponse::new);
+
+    private final Supplier<FrozenCacheService> frozenCacheService;
+
+    @Inject
+    public TransportSearchableSnapshotsNodeCachesStatsAction(
+        ThreadPool threadPool,
+        ClusterService clusterService,
+        TransportService transportService,
+        ActionFilters actionFilters,
+        SearchableSnapshots.FrozenCacheServiceSupplier frozenCacheService
+    ) {
+        super(
+            ACTION_NAME,
+            threadPool,
+            clusterService,
+            transportService,
+            actionFilters,
+            NodesRequest::new,
+            NodeRequest::new,
+            ThreadPool.Names.MANAGEMENT,
+            ThreadPool.Names.SAME,
+            NodeCachesStatsResponse.class
+        );
+        this.frozenCacheService = frozenCacheService;
+    }
+
+    @Override
+    protected NodesCachesStatsResponse newResponse(
+        NodesRequest request,
+        List<NodeCachesStatsResponse> responses,
+        List<FailedNodeException> failures
+    ) {
+        return new NodesCachesStatsResponse(clusterService.getClusterName(), responses, failures);
+    }
+
+    @Override
+    protected NodeRequest newNodeRequest(NodesRequest request) {
+        return new NodeRequest();
+    }
+
+    @Override
+    protected NodeCachesStatsResponse newNodeResponse(StreamInput in) throws IOException {
+        return new NodeCachesStatsResponse(in);
+    }
+
+    @Override
+    protected void resolveRequest(NodesRequest request, ClusterState clusterState) {
+        final ImmutableOpenMap<String, DiscoveryNode> dataNodes = clusterState.getNodes().getDataNodes();
+
+        final DiscoveryNode[] resolvedNodes;
+        if (request.nodesIds() == null || request.nodesIds().length == 0) {
+            resolvedNodes = dataNodes.values().toArray(DiscoveryNode.class);
+        } else {
+            resolvedNodes = Arrays.stream(request.nodesIds())
+                .filter(dataNodes::containsKey)
+                .map(dataNodes::get)
+                .collect(Collectors.toList())
+                .toArray(DiscoveryNode[]::new);
+        }
+        request.setConcreteNodes(resolvedNodes);
+    }
+
+    @Override
+    protected NodeCachesStatsResponse nodeOperation(NodeRequest request, Task task) {
+        final FrozenCacheService.Stats frozenCacheStats = frozenCacheService.get() != null ? frozenCacheService.get().getStats() : null;
+        return new NodeCachesStatsResponse(
+            clusterService.localNode(),
+            frozenCacheStats != null ? frozenCacheStats.getNumberOfRegions() : 0,
```

Comment:
maybe there's a dummy `FrozenCacheService.Stats` object that we can use instead of doing all these null checks here?

Reply:
Sure, I pushed 6e901e8292a70901a2b575a73091815345b3a54e